Uploads per-test Xen console logs to OpenQA

deeplow opened the pull request:
Should only be merged after https://github.com/QubesOS/openqa-tests-qubesos/pull/16.

Logs are obtained by filtering log files based on the start and end time of each test.

deeplow commented:
I should note that I didn't get to test this on openQA side so I have no idea if that is rendered. One hypothesis I have is that the timestamps could should up repeatedly when a test has multiple failures?

marmarek commented on tests/system_tests.pm:
```diff
@@ -100,6 +103,15 @@ ENDFUNC
+        $test_logs_path = "/tmp/$test/";
+        assert_script_run("mkdir $test_logs_path");
+        assert_script_run("sudo python3 split_logs.py --junit-xml=nose2-junit-$test.xml --outdir=$test_logs_path");
+        my @files = <"$test_logs_path*">;
```
This won't work, because the the perl script is running outside of the VM. There is a `script_output` function to run something inside the system under test and return its output.

deeplow replied:
Ah. Makes sense. See if commit 167120b7b1c114cdd000b19a7b93750340aaf5a4 fixes it.

marmarek commented on tests/system_tests.pm:
```diff
@@ -100,6 +103,15 @@ ENDFUNC
         parse_extra_log('JUnit', "nose2-junit-$test.xml");
+
+        # upload per-test logs
+        $test_logs_path = "/tmp/$test/";
```
```suggestion
        my $test_logs_path = "/tmp/$test/";
```

marmarek commented on extra-files/split_logs.py:
```diff
@@ -0,0 +1,90 @@
+                        help='Where to output the split logs.')
+
+    args = parser.parse_args()
+    xml = minidom.parse(args.junit_xml)
```
This would be too easy... The junit XML (at least as generated by nose2) is in fact multiple concatenated documents (multiple root elements).

https://openqa.qubes-os.org/tests/27001#step/system_tests/39

deeplow replied:
Malformed XML :'( I was testing with a local XML I had generated which had only one test suite. I've now tested with [a real one](https://openqa.qubes-os.org/tests/26999).

Rebased and fixed this. But it's still not ready for review. I found another issue.

marmarek commented:
BTW, you need also to copy `split_logs.py` to the home dir - see https://github.com/QubesOS/openqa-tests-qubesos/blob/master/tests/system_tests.pm#L32

deeplow commented:
Rebased and addressed the feedback but I've detected a potential issue.

@marmarek when you have the chance, run a test with these latest changes. it should not break openQA. I need to find a completed test with some failures to analyses the potential issue I've found (has to do with time differences between the timestamp and the times in the logs -- may be shifted by some hours).

marmarek commented on extra-files/split_logs.py:
```diff
@@ -0,0 +1,101 @@
+                        help='Where to output the split logs.')
+
+    args = parser.parse_args()
+    os.mkdir(args.outdir)
```
```suggestion
    os.makedirs(args.outdir, exist_ok=True)
```

deeplow replied:
Rebased and added this suggestion.

marmarek commented:
I have a weird case: https://openqa.qubes-os.org/tests/27103 - during the tests xterm where it was running disappeared. The same happened again when I restarted the test. No idea if it's related to this PR, but there were very few other changes.

deeplow commented:
> I have a weird case: https://openqa.qubes-os.org/tests/27103 - during the tests xterm where it was running disappeared. The same happened again when I restarted the test. No idea if it's related to this PR, but there were very few other changes.

Quite strange. I was taking a look at the logs and nothing jumped out other than seeing in the video the xterm just disappear and the test logs stopping right after `qubes.tests.integ.grub.TC_42_PVHGrub_whonix-ws-16.test_010_template_based_vm` at [3:40min in](https://openqa.qubes-os.org/tests/27103/video?filename=video.ogv&t=220)

marmarek commented on extra-files/split_logs.py:
```diff
@@ -0,0 +1,101 @@
+    for file_name in os.listdir(path):
+        if not file_name.startswith("guest-test-"):
+            continue
+        with open(path + file_name, 'r') as f:
```
```suggestion
        with open(os.path.join(path, file_name), 'r') as f:
```
or at least don't forget the `/` between...

marmarek replied:
This is still an issue

deeplow replied:
fixed

marmarek commented on extra-files/split_logs.py:
```diff
@@ -0,0 +1,101 @@
+            return False
+        else:
+            timestamp = datetime.datetime.fromisoformat(
+                timestamp_match.group(0))
```
Console logs (and other system logs in general) have local timestamps, while the `start_time`/`end_time` are in UTC. You need to convert it here with `.astimezone(datetime.timezone.utc)`.

deeplow replied:
Good point, however first I need to know the timezone the tests VM was on, correct? I was trying to find a way to know that just form the logs or openQA. Does openQA make that possible or do you know of some other way?

marmarek replied:
The `astimezone` method converts the time from _local_ system time to selected one (here UTC), so it should work. When tested manually on one test it gives correct results.

deeplow replied:
addressed in https://github.com/QubesOS/openqa-tests-qubesos/pull/17/commits/46f1fcc51d7cfe580c1bf50cfd3f816a1df85280

marmarek commented on extra-files/split_logs.py:
```diff
@@ -0,0 +1,101 @@
+            time = datetime.timedelta(
+                seconds=float(testcase.attributes['time'].value))
+            timestamp_str = testcase.attributes['timestamp'].value
+            start_time = datetime.datetime.fromisoformat(timestamp_str)
```
Attach UTC timezone to it, to be comparable with timezone aware objects (see above). For example with `.replace(tzinfo=datetime.timezone.utc)`.

deeplow replied:
addressed in https://github.com/QubesOS/openqa-tests-qubesos/pull/17/commits/46f1fcc51d7cfe580c1bf50cfd3f816a1df85280

marmarek commented:
With the above changes, splitting the logs seems to work. But uploading does not: https://openqa.qubes-os.org/tests/27110#step/system_tests/43

marmarek commented:
Hmm, [here](https://openqa.qubes-os.org/tests/27125#step/system_tests/52) splitting did not work :(

deeplow commented:
> But uploading does not: https://openqa.qubes-os.org/tests/27110#step/system_tests/43

https://github.com/QubesOS/openqa-tests-qubesos/pull/17/commits/921c0fb3a2d9bb14a76a99824b48dc89cb99ffc5 might fix it.




> Hmm, [here](https://openqa.qubes-os.org/tests/27125#step/system_tests/52) splitting did not work :(

Maybe it could be related to timezones. I'll investigate.

deeplow commented:
> I have a weird case: https://openqa.qubes-os.org/tests/27103 - during the tests xterm where it was running disappeared. The same happened again when I restarted the test. No idea if it's related to this PR, but there were very few other changes.

Quite strange. I was taking a look at the logs and nothing jumped out other than seeing in the video the xterm just disappear and the test logs stopping right after `qubes.tests.integ.grub.TC_42_PVHGrub_whonix-ws-16.test_010_template_based_vm` at [3:40min in](https://openqa.qubes-os.org/tests/27103/video?filename=video.ogv&t=220)

marmarek commented on tests/system_tests.pm:
```diff
@@ -100,6 +106,18 @@ ENDFUNC
+        my $test_logs_path = "/tmp/$test/";
+        assert_script_run("mkdir $test_logs_path");
+        assert_script_run("sudo python3 split_logs.py --junit-xml=nose2-junit-$test.xml --outdir=$test_logs_path");
+        my $files_path_str = script_output("find $test_logs_path");
```
add `-type f` ?

deeplow replied:
Sure :)

deeplow commented:
> Hmm, [here](https://openqa.qubes-os.org/tests/27125#step/system_tests/52) splitting did not work :(

It seems it did do what was intended after all. What happened is that the test was skipped, but for some reason had cleanup issues. Do you wish the splitting to happen to every log no matter the exit status? If so, just let me know.

deeplow commented:
> > Hmm, [here](https://openqa.qubes-os.org/tests/27125#step/system_tests/52) splitting did not work :(
> 
> ~It seems it did do what was intended after all. What happened is that the test was skipped (see [source here](https://github.com/QubesOS/openqa-tests-qubesos/pull/17/files#diff-0c7b29127253abec05c7aed9cd101a09546ccb8eed7afb05a9c046c201a469c7R81)), but for some reason had cleanup issues. Do you wish the splitting to happen to every log no matter the exit status? If so, just let me know.~
> 
> (spoke too soon. Investigating)

There are two reasons why there are no logs:
  1. The test that failed and thus would have logs is skipped, which means no VM ever starts.
  2. There is that timestamp mismatch between the UTC timestamp in junitxml and the syslogs which are not in UTC (which we talked about earlier)

deeplow commented:
> * There is that timestamp mismatch between the UTC timestamp in junitxml and the syslogs which are not in UTC (which we talked about earlier)

@marmarek I think I've nailed down the timezone conversions. I ended up extracting dom0 timezone from the `systemctl` logs and comparing that with the junitxml which is in UTC. Now the logs should start showing whenever there is a `failure` or `error`.

marmarek reviewed (approved):
It works now!

deeplow commented:
> It works now!

Perfect! Rebased on master (that that the dependency is merged).
